draft-duke-quic-retry-offload.md

```diff
@@ -0,0 +1,876 @@
+list MUST be a Deny-List that is the intersection of all provided Deny-Lists and
+the inverses of all Allow-Lists.
+
+If no entities that control servers have reported Allow-Lists or Deny-Lists,
+the default is a Deny-List with the null set (i.e., all unsupported versions
+will be admitted). This preserves the future extensibilty of QUIC.
+
+A Retry Offload MUST forward all packets for a QUIC version it does not
+support that are not on a Deny-List or absent from an Allow-List. Note that if
+servers support versions the Retry Offload does not, this may increase load on
+the servers.
+
+Note that future versions of QUIC might not have Retry packets, require
+different information in Retry, or use different packet type indicators.
+
+## Considerations for Non-Initial Packets
+
+Initial Packets are especially effective at consuming server resources
+because they cause the server to create connection state. Even when mitigating
+this load with Retry Packets, the act of validating an Initial Token and sending
+a Retry Packet is more expensive than the response to a non-Initial packet with
+an unknown Connection ID: simply dropping it and/or sending a Stateless Reset.
+
+Nevertheless, a Retry Offload in Active Mode might desire to shield servers
+from non-Initial packets that do not correspond to a previously admitted
+Initial Packet. This has a number of considerations.
+
+* If a Retry Offload maintains no per-flow state, it cannot distinguish between
+valid and invalid non-Initial packets and MUST forward all non-Initial Packets
+to the server.
+
+* For QUIC versions the Retry Offload does not support and are present on the
+Allow-List (or absent from the Deny-List), the Retry Offload cannot distinguish
+Initial Packets from other long headers and therefore MUST admit all long
+headers.
+
+* If a Retry Offload keeps per-flow state, it can identify 4-tuples that have
+been previously approved, admit non-Initial packets from those flows, and
+drop all others. However, dropping short headers will effectively break Address
+Migration and NAT Rebinding when in Active Mode, as post-migration packets will
+arrive with a previously unknown 4-tuple. This policy will also break connection
+attempts using any new QUIC versions that begin connections with a short header.
+
+* If a Retry Offload is integrated with a QUIC-LB routable load balancer
+{{?I-D.ietf-quic-load-balancers}}, it can verify that the Destination Connection
+ID is routable, and only admit non-Initial packets with routable DCIDs. As the
+Connection ID encoding is invariant across QUIC versions, the Retry Offload can
+do this for all short headers.
+
+Nothing in this section prevents Retry Offloads from making basic syntax
+correctness checks on packets with QUIC versions that it understands (e.g.,
+enforcing the Initial Packet datagram size minimum in version 1) and
+dropping packets that are not routable with the QUIC specification.
```

Comment:
I'm somewhat worried about this statement. There could be some other input (i.e. resumption token, transport parameter) that breaks some assumption here, even for QUIC v1. I would feel better recommending the Retry Offload only validate the minimum amount necessary to operate correctly.

Reply:
I eliminated "not routable" part because I'm not sure what that meant, and certainly not in this draft.

I don't really understand your concern (how could tokens bring it below the minimum packet size?) but as this is orthogonal to the PR, please file an issue if you'd like to pursue it.